File: scripts/test-serial-interactive.js
#!/usr/bin/env node

/**
 * Interactive serial test - send commands and receive data
 * Run with: node test-serial-interactive.js [baudRate]
 */

const { SerialPort } = require('serialport')
const { ReadlineParser } = require('@serialport/parser-readline')
const readline = require('readline')

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
})

async function interactiveTest(devicePath, baudRate = 9600) {
  console.log(`\n🔌 Connecting to ${devicePath} at ${baudRate} baud...`)
  
  const port = new SerialPort({
    path: devicePath,
    baudRate: baudRate
  })
  
  const parser = new ReadlineParser({ delimiter: '\n' })
  port.pipe(parser)
  
  let messageCount = 0
  
  port.on('open', () => {
    console.log('✅ Connected!\n')
    console.log('Commands you can try:')
    console.log('  - Type any text and press Enter to send')
    console.log('  - Type "exit" to quit')
    console.log('  - Try common commands: "status", "info", "help", "?\n')
    prompt()
  })
  
  parser.on('data', (data) => {
    messageCount++
    const timestamp = new Date().toLocaleTimeString()
    console.log(`\n[${timestamp}] << ${data.trim()}`)
    prompt()
  })
  
  port.on('error', (err) => {
    console.error('❌ Error:', err.message)
  })
  
  function prompt() {
    rl.question('> ', (input) => {
      const command = input.trim()
      
      if (command.toLowerCase() === 'exit') {
        console.log('\n👋 Closing connection...')
        port.close(() => {
          console.log(`Received ${messageCount} messages total.`)
          process.exit(0)
        })
        return
      }
      
      if (command) {
        port.write(command + '\n', (err) => {
          if (err) {
            console.error('❌ Write error:', err.message)
          } else {
            console.log(`>> Sent: ${command}`)
          }
          prompt()
        })
      } else {
        prompt()
      }
    })
  }
}

async function main() {
  console.log('🔧 Interactive Serial Test')
  console.log('===========================')
  
  const { SerialPort } = require('serialport')
  const ports = await SerialPort.list()
  
  const usbPorts = ports.filter(p => 
    p.path.includes('ACM') || 
    p.path.includes('USB') || 
    p.path.includes('ttyUSB') || 
    p.path.includes('ttyACM') ||
    p.path.includes('COM')
  )
  
  if (usbPorts.length === 0) {
    console.log('❌ No USB serial devices found')
    process.exit(1)
  }
  
  console.log('\nAvailable USB devices:')
  usbPorts.forEach((p, i) => {
    console.log(`  [${i}] ${p.path}${p.manufacturer ? ' - ' + p.manufacturer : ''}`)
  })
  
  const testPort = usbPorts[0].path
  const baudRate = process.argv[2] ? parseInt(process.argv[2]) : 9600
  
  console.log(`\nUsing: ${testPort}`)
  
  await interactiveTest(testPort, baudRate)
}

main().catch(err => {
  console.error('💥 Error:', err.message)
  process.exit(1)
})

File: app/scripts/test-serial-interactive.py
#!/usr/bin/env python3
"""
Interactive serial test - send commands and receive data
Run with: python test_serial_interactive.py [baudRate]
"""

import sys
import serial
import serial.tools.list_ports
import threading
from datetime import datetime

def list_usb_devices():
    ports = serial.tools.list_ports.comports()
    usb_ports = [
        p for p in ports
        if ("USB" in p.device or 
            "ACM" in p.device or 
            "ttyUSB" in p.device or 
            "ttyACM" in p.device or
            "COM" in p.device)
    ]
    return usb_ports


def reader_thread(port):
    """Background thread that prints incoming serial data."""
    while True:
        try:
            line = port.readline().decode(errors="replace").strip()
            if line:
                timestamp = datetime.now().strftime("%H:%M:%S")
                print(f"\n[{timestamp}] << {line}")
                print("> ", end="", flush=True)
        except Exception as e:
            print(f"\n❌ Read error: {e}")
            break


def interactive_test(device_path, baud_rate=9600):
    print(f"\n🔌 Connecting to {device_path} at {baud_rate} baud...")
    
    try:
        port = serial.Serial(device_path, baud_rate, timeout=0.1)
    except Exception as e:
        print(f"❌ Failed to connect: {e}")
        sys.exit(1)

    print("✅ Connected!\n")
    print("Commands you can try:")
    print("  - Type any text and press Enter to send")
    print("  - Type 'exit' to quit")
    print("  - Try typical commands: 'status', 'info', 'help', '?'")
    print()

    # Start background thread for reading
    threading.Thread(target=reader_thread, args=(port,), daemon=True).start()

    msg_count = 0

    while True:
        try:
            cmd = input("> ").strip()
        except (EOFError, KeyboardInterrupt):
            cmd = "exit"

        if cmd.lower() == "exit":
            print("\n👋 Closing connection...")
            port.close()
            print(f"Total messages received: {msg_count}")
            break

        if cmd:
            try:
                port.write((cmd + "\n").encode())
                print(f">> Sent: {cmd}")
            except Exception as e:
                print(f"❌ Write error: {e}")


def main():
    print("🔧 Interactive Serial Test")
    print("==========================")

    usb_ports = list_usb_devices()
    if not usb_ports:
        print("❌ No USB serial devices found")
        sys.exit(1)

    print("\nAvailable USB devices:")
    for i, p in enumerate(usb_ports):
        name = p.description or ""
        print(f"  [{i}] {p.device} - {name}")

    chosen = 0  # default to first detected device

    device_path = usb_ports[chosen].device
    baud_rate = int(sys.argv[1]) if len(sys.argv) > 1 else 9600

    print(f"\nUsing: {device_path}")
    interactive_test(device_path, baud_rate)


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"💥 Error: {e}")
        sys.exit(1)
